Cannot configure if Trilinos has CUDA support

I cannot configure deal if I compiled Trilinos with CUDA support. During configuration we check that Sacado supports C++11 (see [here](https://github.com/dealii/dealii/blob/master/cmake/modules/FindTRILINOS.cmake#L151)), the problem is that if CUDA is enabled we should use nvcc to try to compile the test but we are using gcc. So the test fails with the following error:
```
In file included from /home/bruno/Documents/software/trilinos_debug/install/include/Sacado_ConfigDefs.h:80:0,
                 from /home/bruno/Documents/software/trilinos_debug/install/include/Sacado_cmath.hpp:34,
                 from src.cxx:2:
/home/bruno/Documents/software/trilinos_debug/install/include/Kokkos_Macros.hpp:182:40: error: ‘__device__’ does not name a type 
   #define KOKKOS_INLINE_FUNCTION       __device__  __host__  inline 
                                        ^  
/home/bruno/Documents/software/trilinos_debug/install/include/Sacado_cmath.hpp:62:3: note: in expansion of macro ‘KOKKOS_INLINE_FUNCTION’   
   KOKKOS_INLINE_FUNCTION 
   ^ 
ninja: build stopped: subcommand failed. 
```
Since the test fails, we conclude that sacado does not support C++11 and the configuration stops with
```
CMake Error at cmake/macros/macro_configure_feature.cmake:112 (MESSAGE):
  

  Could not find the trilinos library!

  The Trilinos installation (found at
  "/home/bruno/Documents/software/trilinos_debug/install")

  is not compatible with the C++ standard selected for

  this compiler.  See the deal.II FAQ page for a solution.
```

Even if we can go through the configuration, we will need all the files that include Sacado (and probably MueLu) to be compiled with nvcc not gcc.  I am not sure how to fix this... I can live with having to install two different versions of Trilinos but that might be a problem for others if they want to use `Tpetra` with cuda support in their own code.

--- a/include/deal.II/base/numbers.h
+++ b/include/deal.II/base/numbers.h
@@ -166,6 +166,15 @@ namespace numbers
   static const double SQRT1_2 = 0.70710678118654752440;
 
   /**
+   * Check whether the given type can be used in CUDA device code.
+   * If not, DEAL_II_CUDA_HOST_DEV needs to be disabled for functions
+   * that use this type.
+   */
+  template <typename Number, typename = void>
+  struct is_cuda_compatible : std::true_type
+  {};
+
+  /**
    * Check whether a value is not a number.
    *
    * This function uses either <code>std::isnan</code>, <code>isnan</code>, or
@@ -361,10 +370,21 @@ namespace numbers
      * general template is chosen for types not equal to std::complex, this
      * function simply returns the square of the given number.
      *
-     * @note This function can also be used in CUDA device code.
+     * @note If the template type can be used in CUDA device code, the same holds true
+     * for this function.
      */
-    static DEAL_II_CUDA_HOST_DEV real_type
-                                 abs_square(const number &x);
+    template <typename Dummy = number>
+    static DEAL_II_CUDA_HOST_DEV
+      typename std::enable_if<std::is_same<Dummy, number>::value &&
+                                is_cuda_compatible<Dummy>::value,
+                              real_type>::type
+      abs_square(const number &x);
+
+    template <typename Dummy = number>
+    static typename std::enable_if<std::is_same<Dummy, number>::value &&
+                                     !is_cuda_compatible<Dummy>::value,
+                                   real_type>::type
+    abs_square(const number &x);
 
     /**
      * Return the absolute value of a number.
@@ -423,11 +443,19 @@ namespace numbers
 
   // --------------- inline and template functions ---------------- //
 
+  template <typename Number>
+  struct is_cuda_compatible<std::complex<Number>, void> : std::false_type
+  {};
+
+
+
   inline bool
   is_nan(const double x)
   {
     return std::isnan(x);
   }
+
+
 
   inline bool
   is_finite(const double x)
@@ -475,7 +503,23 @@ namespace numbers
 
 
   template <typename number>
-  DEAL_II_CUDA_HOST_DEV typename NumberTraits<number>::real_type
+  template <typename Dummy>
+  DEAL_II_CUDA_HOST_DEV
+    typename std::enable_if<std::is_same<Dummy, number>::value &&
+                              is_cuda_compatible<Dummy>::value,
+                            typename NumberTraits<number>::real_type>::type
+    NumberTraits<number>::abs_square(const number &x)
+  {
+    return x * x;
+  }
+
+
+
+  template <typename number>
+  template <typename Dummy>
+  typename std::enable_if<std::is_same<Dummy, number>::value &&
+                            !is_cuda_compatible<Dummy>::value,
+                          typename NumberTraits<number>::real_type>::type
   NumberTraits<number>::abs_square(const number &x)
   {
     return x * x;

--- a/include/deal.II/base/numbers.h
+++ b/include/deal.II/base/numbers.h
@@ -175,6 +175,13 @@ namespace numbers
   {};
 
   /**
+   * std::complex cannot be used in CUDA device code.
+   */
+  template <typename Number>
+  struct is_cuda_compatible<std::complex<Number>, void> : std::false_type
+  {};
+
+  /**
    * Check whether a value is not a number.
    *
    * This function uses either <code>std::isnan</code>, <code>isnan</code>, or
@@ -442,12 +449,6 @@ namespace numbers
   };
 
   // --------------- inline and template functions ---------------- //
-
-  template <typename Number>
-  struct is_cuda_compatible<std::complex<Number>, void> : std::false_type
-  {};
-
-
 
   inline bool
   is_nan(const double x)

--- a/include/deal.II/differentiation/ad/sacado_number_types.h
+++ b/include/deal.II/differentiation/ad/sacado_number_types.h
@@ -926,6 +926,17 @@ namespace Differentiation
   } // namespace AD
 } // namespace Differentiation
 
+
+namespace numbers
+{
+  template <typename NumberType>
+  struct is_cuda_compatible<
+    NumberType,
+    typename std::enable_if<dealii::Differentiation::AD::is_sacado_rad_number<
+      NumberType>::value>::type> : std::false_type
+  {};
+} // namespace numbers
+
 #  endif // DOXYGEN
 
 

--- a/source/base/tensor.cc
+++ b/source/base/tensor.cc
@@ -19,6 +19,7 @@
 #include <deal.II/base/tensor.h>
 
 #include <deal.II/differentiation/ad/adolc_product_types.h>
+#include <deal.II/differentiation/ad/sacado_number_types.h>
 #include <deal.II/differentiation/ad/sacado_product_types.h>
 
 DEAL_II_NAMESPACE_OPEN

